### PR TITLE
fix(components/forms)!: show all first class errors on touched

### PR DESCRIPTION
**Pull request by Blackbaud-SandhyaRajasabeson**
[AB#2995433](https://dev.azure.com/blackbaud/Products/_boards/board/t/SKY%20UX%20Program/Stories/?workitem=2995433)

BREAKING CHANGE: All first-class errors display when a form control is `touched`. The `maxlength` and file attachment errors display when a form control is `touched` or `dirty`.

### CODE REVIEW

**Comment by blackbaud-sky-build-user** [comment]
[Storybook preview](https://blackbaud.github.io/skyux-pr-preview/2624/storybook/)

Component Storybooks:

- [action-bars-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/action-bars-storybook/)
- [ag-grid-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/ag-grid-storybook/)
- [angular-tree-component-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/angular-tree-component-storybook/)
- [avatar-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/avatar-storybook/)
- [colorpicker-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/colorpicker-storybook/)
- [data-manager-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/data-manager-storybook/)
- [datetime-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/datetime-storybook/)
- [errors-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/errors-storybook/)
- [forms-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/forms-storybook/)
- [indicators-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/indicators-storybook/)
- [inline-form-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/inline-form-storybook/)
- [layout-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/layout-storybook/)
- [lists-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/lists-storybook/)
- [lookup-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/lookup-storybook/)
- [pages-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/pages-storybook/)
- [phone-field-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/phone-field-storybook/)
- [split-view-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/split-view-storybook/)
- [tabs-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/tabs-storybook/)
- [text-editor-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/text-editor-storybook/)
- [theme-storybook](https://blackbaud.github.io/skyux-pr-preview/2624/storybooks/theme-storybook/)

Apps:

- [code-examples](https://blackbaud.github.io/skyux-pr-preview/2624/code-examples/)
- [integration](https://blackbaud.github.io/skyux-pr-preview/2624/integration/)
- [playground](https://blackbaud.github.io/skyux-pr-preview/2624/playground/)

File: libs/components/colorpicker/src/lib/modules/colorpicker/colorpicker.component.html [hunk 1]
```diff
@@ -359,6 +359,8 @@
   [errors]="ngControl?.errors"
   [labelText]="labelText"
   [showErrors]="ngControl?.touched || ngControl?.dirty"
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/src/lib/modules/colorpicker/colorpicker.component.html]: Can we remove the `showErrors` input altogether since we're passing in `dirty` and `touched` directly?

File: libs/components/forms/src/lib/modules/form-error/form-errors.component.ts [hunk 2]
```diff
@@ -49,6 +49,18 @@ export class SkyFormErrorsComponent {
   @Input({ transform: coerceBooleanProperty })
   public showErrors = true;
 
+  /**
+   * Indicates whether the parent component's control is touched
+   */
+  @Input({ transform: coerceBooleanProperty })
```
Comment (Blackbaud-SteveBrush) [inline libs/components/forms/src/lib/modules/form-error/form-errors.component.ts]: Let's use `booleanAttribute` instead of `coerceBooleanProperty`, since it's a core Angular feature.

File: libs/components/forms/src/lib/modules/form-error/form-errors.component.html [hunk 3]
```diff
@@ -1,104 +1,135 @@
-<div *ngIf="labelText && showErrors">
-  <ng-container *ngIf="errors">
-    <sky-form-error
-      *ngIf="errors['required']"
-      errorName="required"
-      [errorText]="'skyux_form_error_required' | skyLibResources: labelText"
-    />
+@if (labelText && showErrors) {
+  <div class="sky-form-errors">
```
Comment (Blackbaud-SteveBrush) [inline libs/components/forms/src/lib/modules/form-error/form-errors.component.html]: Do we need this `div`? It's preferrable to remove extraneous elements, and would allow us to simplify the logic of this file a bit.
```
@if (labelText) {
	@if (touched) {
		@if (errors?.['required']) {
			<sky-form-error />
		}
		
		<!-- Other SKY UX errors... -->
	}
	
	@if (touched || dirty) {
		@if (errors?.['maxlength']; as maxLengthError) {
			<sky-form-error />
		}
		
		<ng-content />
	}
}
```

File: libs/components/forms/src/lib/modules/file-attachment/file-drop/file-drop.component.html [hunk 4]
```diff
@@ -169,11 +169,14 @@
...
   </div>
 </fieldset>
+<!-- TODO: once we add control value accessor, populate the below `sky-form-errors` with the appropriate values for `showErrors`, `touched`, and `dirty`-->
+<!-- for now we are marking touched as true so that file drop will continue to show `required` error-->
 <sky-form-errors
   [class.sky-file-drop-errors]="labelText && rejectedFiles.length"
   [id]="errorId"
   [labelText]="labelText"
   [showErrors]="rejectedFiles.length"
+  [touched]="true"
```
Comment (Blackbaud-SteveBrush) [inline libs/components/forms/src/lib/modules/file-attachment/file-drop/file-drop.component.html]: Could you assign `touched` to `rejectedFiles.length > 0` instead?

File: libs/components/forms/src/lib/modules/file-attachment/file-drop/file-drop.component.html [hunk 5]
```diff
@@ -169,11 +169,14 @@
     </div>
   </div>
 </fieldset>
+<!-- TODO: once we add control value accessor, populate the below `sky-form-errors` with the appropriate values for `showErrors`, `touched`, and `dirty`-->
+<!-- for now we are marking touched as true so that file drop will continue to show `required` error-->
```
Comment (Blackbaud-SteveBrush) [inline libs/components/forms/src/lib/modules/file-attachment/file-drop/file-drop.component.html]: Let's remove this TODO and create a story instead (or update an existing one). TODOs get lost and never found again.

File: libs/components/forms/src/lib/modules/form-error/form-errors.component.spec.ts [hunk 6]
```diff
@@ -57,9 +61,23 @@ describe('Form errors component', () => {
...
+    expect(formError).toBeVisible();
+
+    // all other first class errors
+    componentInstance.dirty = undefined;
+    componentInstance.touched = true;
+    fixture.detectChanges();
+
     [
       'required',
-      'maxlength',
```
Comment (Blackbaud-SteveBrush) [inline libs/components/forms/src/lib/modules/form-error/form-errors.component.spec.ts:62]: Shouldn't `maxlength` also fire for the "touched" state?

File: libs/components/forms/src/lib/modules/checkbox/checkbox-group.component.spec.ts [hunk 7]
```diff
@@ -179,13 +179,17 @@ describe('Checkbox group component', function () {
       const checkbox = getCheckboxes(fixture)?.[0];
       const checkboxInput = checkbox?.querySelector('input');
 
+      // todo: flex topic about this, is it a bug
```
Comment (Blackbaud-SteveBrush) [inline libs/components/forms/src/lib/modules/checkbox/checkbox-group.component.spec.ts]: Replace this entire `it` block with:
```
it('should validate that a checkbox is selected when required', async () => {
  componentInstance.required = true;
  fixture.detectChanges();

  // Trigger validation and mark all checkboxes as touched.
  fixture.componentInstance.contactMethod.updateValueAndValidity();
  fixture.componentInstance.contactMethod.markAllAsTouched();
  fixture.detectChanges();

  const formError = fixture.nativeElement.querySelector('sky-form-error');
  expect(formError).toBeVisible();
  expect(formError.textContent).toContain('Contact method is required.');
});
```
The problem is related to change detection and unit tests. However, if you trigger validation on the checkbox group's `formGroup` (and not its parent), it should bubble up.

Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/forms/src/lib/modules/checkbox/checkbox-group.component.spec.ts]: thanks for looking into this steve! so change detection wasn't running on manually clicking the checkboxes? (which is the only time it was marked as touched)

File: libs/components/forms/src/lib/modules/form-error/form-errors.component.spec.ts (hunk 6 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/forms/src/lib/modules/form-error/form-errors.component.spec.ts:62]: welp, good catch

File: libs/components/forms/src/lib/modules/file-attachment/file-drop/file-drop.component.html (hunk 5 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/forms/src/lib/modules/file-attachment/file-drop/file-drop.component.html]: ill confirm with kerry if such a story exists, but either way. i prefer to keep this here so that if/when we do get to that story this is addressed as whoever takes that story will see the todo at that point

File: libs/components/forms/src/lib/modules/form-error/form-errors.component.html (hunk 3 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/forms/src/lib/modules/form-error/form-errors.component.html]: good call

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.html [hunk 8]
```diff
@@ -38,11 +38,10 @@
           }}
         </option>
       </select>
-    </sky-input-box>
-    <sky-form-errors [labelText]="labelText">
       <sky-form-error
         *ngIf="
-          hostControl?.errors?.['skyDateRange']?.errors.endDateBeforeStartDate
+          hostControl?.errors?.['skyDateRange']?.errors
+            .endDateBeforeStartDate && labelText
```
Comment (Blackbaud-SteveBrush) [inline libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.html:44]: Is `&& labelText` needed since it's being wrapped in the input box now?

File: libs/components/forms/src/lib/modules/file-attachment/file-drop/file-drop.component.html (hunk 5 above)
Comment (Blackbaud-SteveBrush) [reply libs/components/forms/src/lib/modules/file-attachment/file-drop/file-drop.component.html]: Story: https://dev.azure.com/blackbaud/Products/_workitems/edit/3053845 😎 

Another reason we shouldn't leave TODO's in the code: comments are not removed and are published with the bundle, needlessly increasing its size. 
Screenshot 2024-08-22 at 3 17 54 PM

If you absolutely must keep the comment, we should at the very least reword it to not include `showErrors`. 👍

**Review by Blackbaud-SteveBrush — COMMENTED** [review]
The title of this pull request is misleading since the `maxlength` error also shows on `touched`. I would keep the title simple, something like `feat(components/forms)!: display all components form errors on "touched"`, with a more descriptive `BREAKING CHANGE` paragraph in the pull request's description (see [one of Trevor's](https://github.com/blackbaud/skyux/pull/2598) as an example). There, you can discuss the reasoning for the change and why `maxlength` is also shown on "dirty".

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.html (hunk 8 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.html:44]: yup!

**Review by Blackbaud-SteveBrush — DISMISSED** [review]

**Review by Blackbaud-SteveBrush — APPROVED** [review]
